nidmviewerfsl/callSlicer.py

```diff
@@ -0,0 +1,182 @@
+#
+#======================================================================================================
+#
+# Documentation of matrix creation:
+#
+#======================================================================================================
+#
+# When the voxel size in mm in both maps is the same:
+#
+# An SPM map containing a brain volume will be larger than an FSL template holding a brain volume of
+# the same size (more blank space is included at the edges/sides). To resize an SPM map to match the
+# layout of an FSL template the FSL function `flirt` can be used. However, to do this specific transform
+# a resize matrix is required.
+#
+# By default, if resizing a smaller map to a larger map `flirt` creates an empty map of the same size
+# as the larger map and places the smaller map in the front-left hand corner, centered in the z-axis.
+#
+# In other words when the smaller map is enlarged, (x, y, z) in the original small map becomes
+# (x, y, {l_z-s_z}/2 + z) in the larger map where l_z is the z dimension of the larger map and s_z
+# is the z dimension of the smaller map.
+#
+# This means the z dimensions of the SPM brain volume is now aligned with the z dimension of the FSL
+# template brain volume but the x and y dimensions are not. To rectify this the following transform
+# matrix must be used in flirt:
+#
+#  / 1 0 0 dx \
+# |  0 1 0 dy  |
+# |  0 0 1 0   |
+#  \ 0 0 0 1  /
+#
+# Where dx = l_x - s_x, the difference in size of the x dimensions.
+# and dy = l_y - s_y, the difference in size of the y dimensions.
+#
+#-------------------------------------------------------------------------------------------------------
+#
+# When the voxel size in mm in both maps is not the same:
+#
+# When the voxel size in mm is not the same a scaling factor must be added. When the SPM map has a
+# voxel size larger than 2, the matrix to scale and align the SPM map to the FSL 2mm template
+# simplifies to:
+#
+#  / 1 0 0 s*dx \
+# |  0 1 0 s*dy  |
+# |  0 0 1  0    |
+#  \ 0 0 0  1   /
+#
+# where s = 1/v_spm where v_spm is the voxel size of the SPM map.
+#
+#-------------------------------------------------------------------------------------------------------
+# Source: https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FLIRT/FAQ
+# Author: Tom Maullin (29/11/2017)
+
+import subprocess
+import os
+import shutil
+import random
+
+def nifDim(niftiFilename, k):
+    #Retrieve the k dimension of a nifti given it's filename using FSL.
+    
+    if k == 'x':
+        arg = 'dim1'
+    elif k == 'y':
+        arg = 'dim2'
+    elif k == 'z':
+        arg = 'dim3'
+    elif k == 'pix':
+        arg = 'pixdim1'
+    else:
+        error('Enter a valid dimension... x, y, z or pix')
+
+    #Make the command
+    getDimString = "fslhd " + niftiFilename + " | cat -v | grep ^" + arg
+    #Run the command
+    process = subprocess.Popen(getDimString, shell=True, stdout=subprocess.PIPE)
+    output = process.communicate()
```

Comment:
Since we have to wait for the output of the call to `fslhd`, could we use `subprocess.call`  rather than `Popen` + `communicate`? And to enable error handling I would even recommend `subprocess.check_call`. @TomMaullin: what do you think?

Reply:
Hi @cmaumet , 

I originally started with subprocess.call but found I could not retrieve output from it. Whenever I have searched the recommended option is popen as subprocess.call actually just calls popen + wait itself and returns only a boolean telling you whether it has finished. I found the subprocess.py source code and looked at what call is doing and this is the function:

~~~~def call(*popenargs, timeout=None, **kwargs):
    """Run command with arguments.  Wait for command to complete or
    timeout, then return the returncode attribute.

    The arguments are the same as for the Popen constructor.  Example:

    retcode = call(["ls", "-l"])
    """
    with Popen(*popenargs, **kwargs) as p:
        try:
            return p.wait(timeout=timeout)
        except:
            p.kill()
            p.wait()
            raise
~~~~

Unless I have misunderstood, I'm not sure subprocess.call can be used for what we want here as the output from the command line is not easily retrievable and even then it is only running popen and wait anyway.

`check_call` works though as it runs call and throws an error if `call` did not run properly. I am adding this in now! 

~~~~
def check_call(*popenargs, **kwargs):
    """Run command with arguments.  Wait for command to complete.  If
    the exit code was zero then return, otherwise raise
    CalledProcessError.  The CalledProcessError object will have the
    return code in the returncode attribute.

    The arguments are the same as for the call function.  Example:

    check_call(["ls", "-l"])
    """
    retcode = call(*popenargs, **kwargs)
    if retcode:
        cmd = kwargs.get("args")
        if cmd is None:
            cmd = popenargs[0]
        raise CalledProcessError(retcode, cmd)
    return 0
~~~~

Reply:
@TomMaullin: thanks I understand better now. Can you also update the code to remove `shell=True` to avoid security issues as mentioned by @vsoch in her [comment](https://github.com/incf-nidash/nidmresults-fslhtml/pull/9/files/321aa6f3f43d353949465c3d27f2feca56c9bcef#r155252663) above?